### DISCUSSION TITLE
generate a client

### DISCUSSION THREAD

**Discussion by till** [cloudnative-pg/cloudnative-pg#1177] [General]
Hi,

was wondering if anyone had successfully generated a k8s clientset for cloudnative-pg?

I've been trying all day, though I have to admit I don't know tooooo much about CRDs etc. work and generally interact. My guess is that currently the types don't define the necessary tags (e.g. `// +genclient`) for this.

This is what I ran:

```
codegen@387a878fde75:/go/src/k8s.io/code-generator$ ./generate-groups.sh all github.com/cloudnative-pg/client github.com/cloudnative-pg/api/ "postgresql.cnpg.io:v1"
Generating deepcopy funcs
Generating clientset for postgresql.cnpg.io:v1 at github.com/cloudnative-pg/client/clientset
Generating listers for postgresql.cnpg.io:v1 at github.com/cloudnative-pg/client/listers
Generating informers for postgresql.cnpg.io:v1 at github.com/cloudnative-pg/client/informers
```

But everything generated is empty.

For context, I am trying to use Go to create setups and also to pull info from a cluster and I can generally make it work with a dynamic (k8s) client, but that's hacky and requires conversion between the types defined in [api/v1](https://github.com/cloudnative-pg/cloudnative-pg/tree/main/api/v1) every time I make a request.

**Comment by sxd** [comment]
Hi @till 

Did you managed to create your code?

This doesn't look like an issue, not yet at least, moving this into a discussion.

**Reply by till** [reply]
No, didn't. I think the necessary tags are missing in the code base.

**Comment by sxd** [comment]
@till the marker `//+genclient` it's actually really old and it's not supported in kubebuilder, so the answer to your question it's that we cannot generate clients, you need to import the cloudnative-pg api and work with the object using the api and that will solve your issue

**Reply by till** [reply]
I found kubernetes-sigs/kubebuilder#1152 — it looks like it was deprecated indeed. But unsure why, I guess one of the many k8s mysteries. 😂 The discussion on there seems rather moot. I imported the operator code and used the dynamic client, but converting the requests/responses between these types and _unstructured_ feels super hacky and I had hoped for something typed all along.
